internal/access/signup.go

```diff
@@ -8,62 +8,153 @@ import (
 	"golang.org/x/crypto/bcrypt"
 
 	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/email"
 	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/internal/server/providers"
 	"github.com/infrahq/infra/uid"
 )
 
+// NewOrgDetails are details about the identity, org, and access key after a sign-up is successful
+type NewOrgDetails struct {
+	Identity     *models.Identity
+	Organization *models.Organization
+	Bearer       string
+}
+
+// SocialSignup stores the information about a sign-up from a successful OIDC authentication
+type SignupSocial struct {
+	IDPAuth     *providers.IdentityProviderAuth
+	RedirectURL string // stored on provider user to use refresh token in the future
+	Provider    *models.Provider
+}
+
+// SocialUser allows a user to sign-up with an email and a password
+type SignupUser struct {
+	Name     string
+	Password string
+}
+
 type SignupDetails struct {
-	Name      string
-	Password  string
+	User      *SignupUser
+	Social    *SignupSocial
 	Org       *models.Organization
 	SubDomain string
 }
 
-// Signup creates a user identity using the supplied name and password and
-// grants the identity "admin" access to Infra.
-func Signup(c *gin.Context, keyExpiresAt time.Time, baseDomain string, details SignupDetails) (*models.Identity, string, error) {
+// Signup creates an organization, and grants an identity "admin" access to Infra.
+func Signup(c *gin.Context, keyExpiresAt time.Time, baseDomain string, details *SignupDetails) (*NewOrgDetails, error) {
+	if details.Social == nil && details.User == nil {
+		return nil, fmt.Errorf("sign-up requires social login details or user details")
+	}
 	rCtx := GetRequestContext(c)
 	db := rCtx.DBTxn
 
 	details.Org.Domain = SanitizedDomain(details.SubDomain, baseDomain)
 
+	switch {
+	case details.User != nil:
+		email, err := email.Domain(details.User.Name)
+		if err != nil {
+			return nil, fmt.Errorf("set allowed domain from user: %w", err)
+		}
+		details.Org.AllowedDomains = []string{email}
```

Comment:
Unless I'm mistaken, this doesn't omit `gmail.com` nor `googlemail.com` as was previously done in `handlers.go`.

Reply:
Fixed, thanks